webhook/core/annotation_validation_webhook.go

```diff
@@ -30,69 +30,80 @@ import (
+			"annotation key", config.ResourceNamePodENI, "annotation value", val)
+		return admission.Denied(
+			fmt.Sprintf("pod cannot be created with %s annotation", config.ResourceNamePodENI))
+	}
+	return admission.Allowed("")
+}
 
-	// Ignore pod that is scheduled on host network.
-	if pod.Spec.HostNetwork {
-		return admission.Allowed("Pod using host network will not have a pod ENI")
+func (a *AnnotationValidator) handleUpdate(req admission.Request) admission.Response {
+	pod := &corev1.Pod{}
+	if err := a.decoder.DecodeRaw(req.Object, pod); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
 	}
+	oldPod := &corev1.Pod{}
+	if err := a.decoder.DecodeRaw(req.OldObject, oldPod); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+	logger := a.Log.WithValues("name", pod.Name, "namespace", pod.Namespace, "uid", pod.UID)
 
-	if podEniJSON, ok := pod.Annotations[config.ResourceNamePodENI]; ok {
-		webhookLog.V(1).Info("Got annotation:", config.ResourceNamePodENI, podEniJSON)
-		if req.Operation == v1beta1.Create {
-			// Check who is setting the annotation
-			if (req.UserInfo.Username != validUserInfo) && (req.UserInfo.Username != newValidUserInfo) {
-				webhookLog.Info("Denying annotation creation", "Username", req.UserInfo.Username)
-				return admission.Denied("Validation failed. Pod ENI not set by VPC Resource Controller")
-			}
-		} else if req.Operation == v1beta1.Update {
-			// Check if the pod-eni annotation has been changed
-			webhookLog.V(1).Info("Operation is update, checking that pod-eni annotation wasn't modified")
-			oldPod, err := av.K8sWrapper.GetPod(pod.Namespace, pod.Name)
-			if err != nil {
-				webhookLog.Error(err, "Failed to fetch pod in update request")
-				return admission.Denied("Validation failed. Trying to update annotation on a pod that can't be found")
-			}
-			if oldPodAnnotationValue, ok := oldPod.Annotations[config.ResourceNamePodENI]; ok {
-				// This is an update trying to change the pod annotation
-				if oldPodAnnotationValue != podEniJSON && (req.UserInfo.Username != validUserInfo) && (req.UserInfo.Username != newValidUserInfo) {
-					webhookLog.Info("Denying annotation change", "Username", req.UserInfo.Username)
-					return admission.Denied("Validation failed. Pod ENI annotation changed outside of VPC Resource Controller")
-				}
-			} else {
-				// This is an update trying to add the pod annotation
-				if (req.UserInfo.Username != validUserInfo) && (req.UserInfo.Username != newValidUserInfo) {
-					webhookLog.Info("Denying adding annotation", "Username", req.UserInfo.Username)
-					return admission.Denied("Validation failed. Pod ENI annotation added outside of VPC Resource Controller")
-				}
-			}
+	// This will block any update on the specific annotation from non vpc resource controller
+	// service accounts
+	if pod.Annotations[config.ResourceNamePodENI] !=
+		oldPod.Annotations[config.ResourceNamePodENI] {
```

Comment:
Just to confirm, if `config.ResourceNamePodENI` does not exist in annotation, is it going to return empty string? so we don't check if it exist or not?

Reply:
This statement is supposed to cover the following update scenarios by a non vpc-resource-controler user.
- If the `config.ResourceNamePodENI` is added/updated by any user it should be blocked.
- If the `config.ResourceNamePodENI` is removed by any user it should be blocked.

We could definitely check if the annotation exists or not. Do you think with the current check we could miss some corner cases?

Reply:
Could this case happen?
1, a pod update event coming
2, the pod has same annotations during the update
3, the pod has different usernames

Reply:
Yes this could happen. If the annotations are same and it's an update event from a different username, it should go through.

I'll add another integration test just to be on safe side.